File: src/templates/gemini.ts
/**
 * src/templates/gemini.ts
 *
 * Google Gemini Native REST API template.
 * Translates OpenAI-compatible requests to Gemini Native (generateContent)
 * and translates Gemini chunks back to OpenAI format.
 *
 * This "solve all the problem" version uses Google's native schema,
 * which is much more robust for thinking/reasoning features.
 */

import type { ProviderTemplate, ThinkingMode } from "./types.js";

const THINKING_LEVEL_MAP: Record<ThinkingMode, string> = {
  high: "high",
  medium: "medium",
  low: "low",
};

/**
 * Extract plain text from OpenAI message content.
 * Content can be a string OR an array of parts like [{ type: "text", text: "..." }].
 */
function extractText(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .filter((p: any) => p.type === "text" && p.text)
      .map((p: any) => p.text)
      .join("\n");
  }
  return "";
}

export const geminiTemplate: ProviderTemplate = {
  name: "gemini",

  /** Redirects to the native Gemini REST endpoint and appends the API key as a query param. */
  targetUrlCallback(baseUrl: string, body: any, apiKey: string): string {
    const model = body.model || "gemini-2.5-flash";
    const action = body.stream ? "streamGenerateContent" : "generateContent";
    // Ensure we don't have double slashes and no /openai suffix
    const base = baseUrl.replace(/\/openai\/?$/, "").replace(/\/$/, "");
    const sse = body.stream ? "&alt=sse" : "";
    return `${base}/models/${model}:${action}?key=${apiKey}${sse}`;
  },

  /** Removes the Bearer token as Gemini native uses the 'key' query parameter. */
  applyHeaders(headers: Record<string, string>): void {
    delete headers["Authorization"];
  },

  /** Translates OpenAI Request -> Gemini Native Request. */
  transformRequest(body: any, options: { mode: ThinkingMode | null; provider?: string | null }): any {
    const messages = body.messages || [];
    
    // 1. Map messages to Gemini 'contents' and 'systemInstruction'
    const contents: any[] = [];
    const sysParts: any[] = [];
    let lastRole = "";

    for (const m of messages) {
      const text = extractText(m.content);

      if (m.role === "system") {
        sysParts.push({ text });
        continue;
      }
      
      const role = m.role === "assistant" ? "model" : "user";

      if (role === lastRole && contents.length > 0) {
        // Merge with last message
        contents[contents.length - 1].parts[0].text += `\n\n${text}`;
      } else {
        contents.push({
          role,
          parts: [{ text }]
        });
        lastRole = role;
      }
    }

    const systemInstruction = sysParts.length > 0 ? { parts: sysParts } : undefined;

    // 2. Build Gemini config
    const generationConfig: Record<string, any> = {
      temperature: body.temperature ?? 0.7,
      maxOutputTokens: body.max_tokens,
      stopSequences: body.stop,
      topP: body.top_p,
    };

    // 3. Handle Thinking
    if (options.mode) {
      const isGemini25 = (body.model || "").includes("2.5");
      
      if (isGemini25) {
        // Gemini 2.5 uses a numeric thinking_budget (tokens)
        const budgets: Record<ThinkingMode, number> = {
          high: 24576,
          medium: 8192,
          low: 1024,
        };
        generationConfig.thinking_config = {
          thinking_budget: budgets[options.mode],
          include_thoughts: true,
        };
      } else {
        // Gemini 3.0+ uses thinking_level strings
        generationConfig.thinking_config = {
          thinking_level: THINKING_LEVEL_MAP[options.mode],
          include_thoughts: true,
        };
      }
    }

    return {
      contents,
      systemInstruction,
      generationConfig,
    };
  },

  /** Translates Gemini Native Chunk -> OpenAI Delta Chunk. */
  transformChunk(chunk: any, metadata: { model?: string; id?: string }): any {
    const candidate = chunk.candidates?.[0];
    if (!candidate) return null;

    const parts = candidate.content?.parts || [];
    let text = "";
    for (const p of parts) {
      if (p.text) text += p.text;
      // Suppress raw "thought" booleans that leak through in some models
      // if (p.thought) text += `\n<thought>\n${p.thought}\n</thought>\n`;
    }

    const isFinished = candidate.finishReason === "STOP" || candidate.finishReason === "MAX_TOKENS";

    // Always emit a chunk if there's text OR if this is the terminal chunk with finish_reason
    if (!text && !isFinished) return null;

    return {
      id: metadata.id,
      object: "chat.completion.chunk",
      created: Math.floor(Date.now() / 1000),
      model: metadata.model,
      choices: [
        {
          index: 0,
          delta: text ? { content: text } : {},
          finish_reason: isFinished ? "stop" : null,
        }
      ]
    };
  },

  /** Translates full Gemini Native Response -> OpenAI Chat Completion (non-streaming). */
  transformResponse(response: any, metadata: { model?: string; id?: string }): any {
    const candidate = response.candidates?.[0];
    if (!candidate) return null;

    const content = candidate.content?.parts?.[0]?.text || "";

    return {
      id: metadata.id,
      object: "chat.completion",
      created: Math.floor(Date.now() / 1000),
      model: metadata.model,
      choices: [
        {
          index: 0,
          message: {
            role: "assistant",
            content,
          },
          finish_reason: candidate.finishReason === "STOP" ? "stop" : null,
        }
      ],
      usage: response.usageMetadata ? {
        prompt_tokens: response.usageMetadata.promptTokenCount,
        completion_tokens: response.usageMetadata.candidatesTokenCount,
        total_tokens: response.usageMetadata.totalTokenCount,
      } : undefined,
    };
  },

  /** Not used when transformRequest is present. */
  applyTemplateConfig() {}
};

File: src/gemini.py
"""
gemini.py — OpenAI <-> Gemini request/response transformation.

Ports the TypeScript transformation logic from server.ts (lines 76-174)
into Python.  Handles conversion of OpenAI chat messages to Gemini's
contents + systemInstruction format and vice-versa for responses.
"""

import time
import uuid


def build_gemini_request_body(body: dict) -> dict:
    """Convert an OpenAI chat-completion body to a Gemini API request body.

    Maps:
    - ``system`` messages → ``systemInstruction``
    - ``user`` messages → contents with role ``"user"``
    - ``assistant`` messages → contents with role ``"model"``
    - Generation parameters (temperature, top_p, top_k, n, max_tokens, stop)
      → ``generationConfig`` with Gemini field names.

    Also supports a legacy ``prompt`` field (string or object with
    ``context`` / ``messages`` keys) for backwards compatibility.
    """
    request_body: dict = {}
    contents: list[dict] = []
    system_parts: list[str] = []
    generation_config: dict = {}

    def push_content(role: str, text: str) -> None:
        contents.append({"role": role, "parts": [{"text": text}]})

    def push_message(message: dict) -> None:
        raw = message.get("content", "")
        text = raw if isinstance(raw, str) else str(raw)

        role = message.get("role", "")
        if role == "system":
            system_parts.append(text)
        elif role == "assistant":
            push_content("model", text)
        else:
            push_content("user", text)

    messages = body.get("messages")
    if isinstance(messages, list):
        for msg in messages:
            push_message(msg)

    prompt = body.get("prompt")
    if isinstance(prompt, str):
        push_content("user", prompt)
    elif isinstance(prompt, dict):
        context = prompt.get("context")
        if isinstance(context, str):
            system_parts.append(context)
        prompt_messages = prompt.get("messages")
        if isinstance(prompt_messages, list):
            for msg in prompt_messages:
                push_message(msg)

    if contents:
        request_body["contents"] = contents

    if system_parts:
        request_body["systemInstruction"] = {
            "parts": [{"text": "\n\n".join(system_parts)}]
        }

    temp = body.get("temperature")
    if isinstance(temp, (int, float)):
        generation_config["temperature"] = temp

    top_p = body.get("top_p")
    if isinstance(top_p, (int, float)):
        generation_config["topP"] = top_p

    top_k = body.get("top_k")
    if isinstance(top_k, (int, float)):
        generation_config["topK"] = top_k

    n_val = body.get("n")
    if isinstance(n_val, (int, float)):
        generation_config["candidateCount"] = n_val

    max_tokens = body.get("max_tokens")
    if isinstance(max_tokens, (int, float)):
        generation_config["maxOutputTokens"] = max_tokens

    stop = body.get("stop")
    if stop is not None:
        generation_config["stopSequences"] = (
            stop if isinstance(stop, list) else [str(stop)]
        )

    if generation_config:
        request_body["generationConfig"] = generation_config

    return request_body


def transform_gemini_response(gemini_response: dict) -> dict:
    """Convert a Gemini API response to an OpenAI chat.completion payload.

    Extracts text from ``candidates[0].content.parts[].text``, maps
    ``usageMetadata`` to token counts, and normalises ``finishReason``.
    """
    usage_meta = gemini_response.get("usageMetadata") or {}

    openai_response: dict = {
        "id": gemini_response.get("responseId") or f"chatcmpl-{uuid.uuid4().hex[:12]}",
        "object": "chat.completion",
        "created": int(time.time()),
        "model": gemini_response.get("modelVersion") or "gemini",
        "choices": [],
        "usage": {
            "prompt_tokens": usage_meta.get("promptTokenCount", 0),
            "completion_tokens": usage_meta.get("candidatesTokenCount", 0),
            "total_tokens": usage_meta.get("totalTokenCount", 0),
        },
    }

    candidates = gemini_response.get("candidates")
    if candidates and len(candidates) > 0:
        candidate = candidates[0]
        content_obj = candidate.get("content") or {}
        parts = content_obj.get("parts") or []
        text = "".join(part.get("text", "") for part in parts if isinstance(part, dict))

        finish_reason = candidate.get("finishReason", "stop")
        if isinstance(finish_reason, str):
            finish_reason = finish_reason.lower()

        openai_response["choices"].append({
            "index": candidate.get("index", 0),
            "message": {
                "role": "assistant",
                "content": text,
            },
            "finish_reason": finish_reason,
        })

    return openai_response
